### PR TITLE
React dialog generic

### PR DESCRIPTION
**Pull request by Bjvanminnen**
Builds on react dialog PR #6166, extracting Dialog html into its own component for easier future reusability. Could make the argument that it makes sense to hold off on doing this until we actually want to reuse it, but I think we will want to. As we do so, we may need/want to increase the complexity of Dialog.jsx

Testing

Did some very basic testing in Chrome

can still close/open dialog multiple times
can close dialog in various ways (click backdrop, press esc, click close button, click x in corner)
Will probably try to run UI tests against this. (passed)

### CODE REVIEW

**Comment by Bjvanminnen** [comment]
This was originally https://github.com/code-dot-org/code-dot-org/pull/6167, but that would have merged into a different branch than staging.

File: dashboard/app/assets/javascripts/components/share_dialog.jsx [hunk 1]
```diff
@@ -21,66 +23,36 @@ window.dashboard.ShareDialog = (function (React) {
...
+    close: function () {
+      this.setState({isOpen: false});
     },
 
     render: function () {
-      if (this.state.hidden) {
-        return <div/>;
-      }
-
       // TODO - Can we now make SendToPhone completely Reactified?
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/components/share_dialog.jsx:38]: Comment still relevant?

Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/components/share_dialog.jsx:38]: I would say yes in this PR. It should be removed in part 3 (https://github.com/code-dot-org/code-dot-org/pull/6171), though it doesn't look like it is there either.

File: dashboard/app/assets/javascripts/components/dialog.jsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    componentDidMount: function () {
+      this.focusDialog();
+    },
+
+    componentDidUpdate: function () {
+      this.focusDialog();
+    },
+
+    closeOnEscape: function (event) {
+      if (event.key === 'Escape') {
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/components/dialog.jsx:26]: I don't think `event.key` has wide browser support?  https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key#Browser_compatibility

Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/components/dialog.jsx:26]: React uses simulated events, which should have cross browser support https://facebook.github.io/react/docs/events.html

Comment (joshlory) [inline dashboard/app/assets/javascripts/components/dialog.jsx:26]: Cool, TIL!

**Comment by islemaster** [comment]
LGTM, very clean!

Don't let it block this PR, but I'd love to see a simple unit test around the new Dialog component.

**Comment by Bjvanminnen** [comment]
One challenge is that I don't think we currently have any good way of unit testing JS that lives in dashboard (this should probably be a red flag for us). I think as part of our JS pipeline work, this component should eventually end up in code-studio, where we should start building a test framework that _can_ test this and other JS

**Pull request merged**